File: 427/Lyubimov Grigoriy/[1].py
import datetime
import queue

# Входные данные
my_data = ("Любимов Григорий Ильич", 4, 1, 2003)
disc_marks = {
    "русский": 5,
    "литература": 5,
    "алгебра": 5,
    "история": 5,
    "обж": 5,
    "геометрия": 5,
    "английский": 5,
    "физика": 5,
    "химия": 5,
    "физра": 5,
    "музыка": 5,
    "труд": 5,
    "обществознание": 5,
    "география": 5,
}
family_names = ["Дима", "Илья", "Елена", "Анна", "Алексей", "Сергей", "Татьяна", "Татьяна", "Полина", "Алина", "Андрей"]
kiwa_name = "Афина"


# 1. средняя оценка
average_mark = sum(disc_marks.values()) / len(disc_marks.values())
print("средняя оценка: ", average_mark)


# 2. уникальные имена среди своих родственников (включая свое)
modified_family_names = ["Григорий"]
unique_names = []
for name in family_names:
    if name not in unique_names:
        unique_names.append(name)
        modified_family_names.append(name)
print("уникальные имена среди родственников: ", *modified_family_names)


# 3-4.общая длина всех названий предметов и общая длина всех названий предметов
len_of_subjects = 0
unique_letters = []
for subject in disc_marks:
    len_of_subjects += len(subject)
    for letter in subject:
        if letter in unique_letters:
            continue
        unique_letters.append(letter)
print("общая длина всех названий предметов: ", len_of_subjects)
print("общая длина всех названий предметов: ", *unique_letters)


# 5. имя в бинарном виде
print("имя в бинарном виде: ")
for ch in bytearray(kiwa_name, 'utf-8'):
    print(bin(ch))


# 6. отсортированный по алфавиту (в обратном порядке) список родственников
unique_names.sort()
unique_names.reverse()
print(*unique_names)


# 7. количество дней от даты моего рождения до текущей даты
my_birthday_date = datetime.date(2003, 1, 4)
current_date = datetime.date.today()
number_of_days = current_date-my_birthday_date
print("количество дней от даты моего рождения:  ", number_of_days.days)


# 8. FIFO очередь
queue = queue.Queue()
while True:
    element = input("Введите элемент очереди(out для вывода) ")
    if element == "out":
        print("Элементы очереди: ")
        while not queue.empty():
            print(queue.get())
        break
    else:
        queue.put(element)


# 9. замена имени в отсортированном списке родственников на имя ацтекского правителя
day = my_data[1]
month = my_data[2]
year = my_data[3]
number_of_aztec = (day + month**2 + year) % 21 + 1
number = int(input("Введите номер для замены имени родственника на имя ацтека: "))
aztec_rulers = ["Huitzilihuitl", "Chimalpopoca", "Xihuitl Temoc", "Acamapichtli", "Tenoch", "Itzcoatl", "Moctezuma I",
                "Atotoztli", "Axayacatl", "Tizoc", "Ahuitzotl", "Moctezuma II", "Cuitláhuac", "Cuauhtémoc",
                "Motelchiuhtzin", "Xochiquentzin", "Huanitzin", "Tehuetzquititzin", "Cecetzin", "Cipac"]
unique_names[number] = aztec_rulers[number_of_aztec]
print(*unique_names)


# 10. связный список
linked_family_list = {"Алексей": 1, "Сергей": 2, "Татьяна": 3, "Елена": 4, "Илья": 5, "Анна": 6, "Григорий": 7,
                      "Дима": 8, "Алина": 9, "Полина": None}

print(linked_family_list)


# 11. Генератор
number_of_generator = (len(my_data[0]) * len(family_names)) % 4
print("Номер генератора: ", number_of_generator)
N = int(input("Введите кол-во чисел трибоначчи(начиная с 3-его): "))


def tribonacci(n):
    t1 = 0
    t2 = 0
    t3 = 1
    for a in range(n):
        c = t1+t2+t3
        t1 = t2
        t2 = t3
        t3 = c
        yield c


tri = tribonacci(N)
tribo = [0, 0, 1]
for i in tri:
    tribo.append(i)
print(tribo)

File: 429/Utrosin/[1]/[1].cpp
#include <iostream>
#include <tuple>
#include <string>
#include <map>
#include <list>
#include <vector>
#include <algorithm>
#include <set>
#include <bitset>
#include <ctime>
#include <queue>
#include < Windows.h >

#pragma warning(disable : 4996)
using namespace std;

int main()
{
	SetConsoleCP(1251);
	SetConsoleOutputCP(1251);
	setlocale(LC_ALL, "Russian");
	//Исходные данные
	// Кортеж
	tuple<string, string, string, string, string, string >Info;
	Info = make_tuple("Утросин", "Кирилл", "Сергеевич", "24", "06", "2004");

	// Словарь
	map<string, unsigned> atestat 
	{
		{"Русский", 4}, {"Матан", 4},{"Литература", 3},{"Физра", 4},
		{"История России", 5},{"Всеобщая История", 5},{"Информатика", 5},
		{"Геометрия", 5},{"Физика", 4},{"Химия", 3},{"Изо", 4},
		{"Астрономия", 4},{"Прогулы", 5}, {"Разговоры о важном", 5}
	};

	// Список родственников
	vector<string> relatives{"Мария 1950","Григорий 1994", "Александр 1975", "Павел 1972",
		"Сергей 1971", "Наташа 1983", "Александр 1956", "Михаил 2002", "Дмитрий 2003", "Дмитрий 2007",
		"Галина 1949", "Алина 1955", "Вера 1999", "Антон 1991", "Слава 2001", "Крипер 2004" };
	
	// Кива
	string kiwa = "Мявка";

#pragma region Задание №1
	int avg_mark = 0;
	for (const auto& pair : atestat) 
	{
		avg_mark += pair.second;
	}
	avg_mark = avg_mark / atestat.size();
	cout << "Средняя оценка по аттестату: " << avg_mark << endl;
#pragma endregion
#pragma region Задание №2
	vector<string> rel;
	vector<string> unique_relatives;
	string name;
	string uniq;
	int j = 0;
	cout << "  " << endl;
	for (int i = 0; i < relatives.size(); i++)
	{
		name = relatives[i];
		while (name[j] != ' ') 
		{
			uniq += name[j];
			j += 1;
		}
		rel.push_back(uniq);
		uniq = "";
		j = 0;
	}
	
	uniq = "";
	unique_relatives = rel;

	sort(unique_relatives.begin(), unique_relatives.end());
	unique_relatives.erase(unique(unique_relatives.begin(), unique_relatives.end()), unique_relatives.end());

	for (const auto& x : unique_relatives) 
	{
		cout << x << " ";
	}

	cout << endl;

#pragma endregion
#pragma region Задание №3
	string dlina = "";
	for (const auto& pair : atestat)
	{
		dlina += pair.first;
	}
	cout << "Общая длина всех названий предметов: " << dlina.length() << endl;
#pragma endregion
#pragma region Задание №4
	set<char> unique_letters;
	string letters = "";
	for (const auto& pair : atestat)
	{
		letters += (pair.first);
	}

	cout << "Уникальные буквы в названиях предметов: ";
	for (char letter : letters) {
		unique_letters.insert(letter);
	}
	for (char n : unique_letters)
		cout << n;
	cout << endl;

#pragma endregion
#pragma region Задание №5
	cout << "Имя Мявка в бинарном коде: " << endl;
	for (int i = 0; i < kiwa.size(); i++) {
		cout << bitset<8>(kiwa[i]) << ' ';
	}
	cout << endl;
#pragma endregion
#pragma region Задание №6
	vector<string> sorted_relatives = relatives;
	sort(sorted_relatives.rbegin(), sorted_relatives.rend());

	cout << "Отсортированный в обратном порядке список родственников: " << endl;
	for (const auto& f : sorted_relatives) {
		cout << f << ", ";
	}
	cout << endl;
#pragma endregion
#pragma region Задание №7
	int year = 2003;
	int month = 6;
	int day = 24;

	time_t now = time(0);
	tm* ltm = localtime(&now);

	tm specified_tm = { 0 };
	specified_tm.tm_year = year - 1900;
	specified_tm.tm_mon = month - 1;
	specified_tm.tm_mday = day;

	time_t specified_date = mktime(&specified_tm);
	double diff = difftime(now, specified_date);

	int days = static_cast<int>(diff / (60 * 60 * 24));

	cout << "Разница в днях между др и текущей датой: " << days << endl;


#pragma endregion
#pragma region Задание №8
	queue<string> predm;

	cout << "Введите предметы, 'stop' - остановка:" << endl;

	string input;
	int index = 0;

	while (input != "stop") {
		cin >> input;

		if (input != "stop") {
			predm.push(input);
			cout << "Предмет: '" << input << "' индекс: " << index << endl;
			index++;
		}
	}

	cout << "Введенные предметы:" << endl;

	while (!predm.empty()) {
		cout << predm.front() << endl;
		predm.pop();
	}
#pragma endregion
#pragma region Задание №9
	vector<string> Actec{"Tenoch","Acamapichtli", "Huitzilihuitl", "Chimalpopoca",
		"Xihuitl Temoc", "Itzcoatl", "Moctezuma I", "Atotoztli", "Axayacatl",
		"Tizoc","Ahuitzotl","Moctezuma II", "Cuitláhuac",
		"Cuauhtémoc", "Tlacotzin", "Motelchiuhtzin", "Xochiquentzin",
		"Huanitzin", "Tehuetzquititzin", "Cecetzin", "Cipac" };
	int number = (day + int(pow(month, 2)) + year) % 21 + 1; //6
	//cout << endl << number;
	
	int Index;
	string Aname;
	cout << "Кого хотите переименовать?: ";
	cin >> Index;
	if (Index >= 0 && Index < sorted_relatives.size()) {
		Aname = Actec[5] + " " + sorted_relatives[index].substr(sorted_relatives[Index].find(' ') + 1);
		sorted_relatives[Index] = Aname;
		cout << "Имя " << sorted_relatives[Index] << " было изменено" << endl;
	}
	else {
		cout << "Индекс вне диапазона" << endl;
	}

#pragma endregion

}
